### ISSUE TITLE
Invalid location with utf-8 encoded files

### ISSUE THREAD

**Issue by hhugo** [ocaml/ocaml#11899]
(Not really a bug since the compiler doesn't officially support utf-8 encoded sources. But ..)

Reading https://github.com/ocaml/ocaml/pull/11736 and working on utf support in jsoo made me realize that while one can already use utf-8 in string literal and comments, the computed locations will be wrong. Computed column would be number of bytes instead of number codepoint.

I didn't read #11736 carefully but I suspect it would also result in bad locations.

Here is a small example:

We currently have
```
$ cat c.ml
let (s : string) = "npiπππ" ^ 2
$ ocamlc -c c.ml
File "c.ml", line 1, characters 33-34:
1 | let (s : string) = "npiπππ" ^ 2
                                     ^
Error: This expression has type int but an expression was expected of type
         string
```

where it should be

```
$ cat c.ml
let (s : string) = "npiπππ" ^ 2
$ ocamlc -c c.ml
File "c.ml", line 1, characters 30-31:
1 | let (s : string) = "npiπππ" ^ 2
                                  ^
Error: This expression has type int but an expression was expected of type
         string
```

**Comment by gasche** [comment]
I think we discussed that before (but I can't find a reference back) with @dbuenzli in particular. There are two different aspects:

1. Should the numbers reported count in terms of bytes or codepoints or graphemes or whatever? My understanding here is that most tools are still using "number of bytes" (and assuming UTF8 encoding), so it is in fact probably better for the user experience to keep to "number of bytes" for now. (But: we are slowly getting the tech in the stdlib to count something more high-level, so maybe we could consider that.)
2. Highlighting should be done correctly. Obviously *that* is a bug in the example you are showing here. When we want to highlight a location by reprinting on the line or under it, we should compute the spacing width correctly.

My intuition would be to keep (1) as-is and try to fix (2) for now (not me personally, I would prefer to leave this to someone else). This would suggest rephrasing the issue as "incorrect highlighting" rather than "incorrect location".

**Comment by nojb** [comment]
As mentioned, the number of codepoints is not the right metric either because codepoints do not correspond to graphical characters in general. 

I recommend https://github.com/dbuenzli/uucp/blob/b1b445f1426db277e58a2402350a6b19644cbd6e/src/uucp.mli#L522-L599  to get the basic background on this topic.

**Comment by dbuenzli** [comment]
See also issue #8978 and the linked discussion in `caml-mode`, #6518 and #8541.

I think that for now it's better to keep things as they are.  

More precisely regarding @gasche points:

1. On the editor side it's a mess, line number + offset in scalar values would actually be a good metric to give them (as it's reliable and easy to compute for tools) but something like [gnu's format](https://www.gnu.org/prep/standards/standards.html#index-formatting-error-messages) expects columns as computed by the output of `wcwidth` which is a broken thing to start with. LSP is a bit better but still a bit [absurd](https://github.com/ocaml/ocaml/pull/8541#issuecomment-476537093) (line number + offset in UTF-16 code points).

   As far as I remember ocaml's format is actually quite good (line number + byte offsets on the line) and it is possible to nowadays to configure emacs so that it highlights the right stuff (see the `caml-mode` discussion linked to in #8978).

2. In terminals there's only so much you can do. @nojb gave a link to the results of the extensive research done by @pqwy who has been polishing a best-effort `wcwidth` that he gave to `uucp`. For now I'd say just forget about it.

**Comment by gasche** [comment]
Maybe there is a way to highlight locations that does not require computing widths in this reportedly-fiddly way? For example we could stick to "on-line" highlighting formats that use markup (underline or bold or...) on the string directly. We print the line counting the number of bytes, and insert the proper escape codes / whatever on the location, without making assumptions about the width of everything.

Then for the "dumb" terminal mode where we don't allow ourselves these revolutionary graphical gimmicks, we could have a simple, approximative test "this line only contains width-1 characters" up to and including the location of interest, and disable the highlighting (just quote the source) when it fails?

**Comment by hhugo** [comment]
> the number of codepoints is not the right metric either because codepoints do not correspond to graphical characters in general.

For error messages, I would want locations that correspond to what my editor understand. I use emacs which I think understand codepoints, not graphical characters.

**Comment by hhugo** [comment]
> [gnu's format](https://www.gnu.org/prep/standards/standards.html#index-formatting-error-messages) expects columns as computed by the output of wcwidth which is a broken thing to start with

@dbuenzli, you say it's broken because it depends on the rendered ?

**Comment by dbuenzli** [comment]
> For example we could stick to "on-line" highlighting formats that use markup (underline or bold or...) on the string directly.

That's a good idea.

> @dbuenzli, you say it's broken because it depends on the rendered ?

Yes there's a whole lot of stuff that can visually happen to a given sequence of scalar values when it hits your pixels (for nice examples see the paper mentioned in #10749).

Asking tools to compute visual columns without help from the rendering layer is hopeless. For now the best is to provide a semi-human format that can at least be picked-up reliably by the rendering layer (for which "this range of bytes on that line in the text you are being the steward of" is not a bad idea).

**Comment by Octachron** [comment]
The `wcwidth` function is just generally broken. First, its type does not make sense for abugidas (like Hindi written in devanagari) where code points are not graphical character but possibly one "half" of a character (see 'री' which is two code points but one character). Even for logograms, only CJK scripts are supported. For instance,  `wcwidth` returns `1` for cuneiforms characters, (for instance `𒆍`)  which cannot fit on one column.

I agree that in general letting the graphical interface handles the highlighting on its own by using colors and/or underlining whenever possible is a good idea. 

For location reporting, I think that the context is important: when communicating with other tools using bytes is the most unambiguous standard. For humans, it would make sense to report location using the number of extended grapheme clusters (since it is the notion of characters than a human can reliably count).

**Comment by dbuenzli** [comment]
> For humans, it would make sense to report location using the number of extended grapheme clusters (since it is the notion of characters than a human can reliably count).

Not even, this is still computed on the [logical order](https://unicode.org/glossary/#logical_order). If you want to help humans you need to report it on the [display order](https://unicode.org/glossary/#display_order) (so you'd need to at least run the [bidi algorithm](https://www.unicode.org/reports/tr9/) and pray that the rendering layer agrees with you).

**Comment by Octachron** [comment]
If we implement the "security" measure that we only have right-to-left isolates in string literals and comments, and since compiler errors never point inside comments and string literals, we could consider that the logical order and the display order agree, don't we?

But I agree that this is too many `if`s compared to letting the rendering layer handles its own issues.

**Comment by dbuenzli** [comment]
Well after bidi you still have [shaping](https://unicode.org/glossary/#shaping_characters), so one more `if` would be for your string literals and comments to not contain scripts that have sophisticated shaping rules like arabic :–)

**Comment by Octachron** [comment]
Should we maybe report locations using the number of tokens? At least, this is a well-defined notion.

Unicode trivia of the day: Arabic has word-level ligatures (like `ﷰ` ), `द्ध्र्य` is a much better example for the complexity of Abugida (my terminal and browser wildly disagree on the shaping of this text) , and no one has still any idea on how to render`𓅨𓐴𓐷𓏤𓐱𓏤𓐰𓏤𓐸𓐰𓂋`.

**Comment by gasche** [comment]
I made concrete proposals for how to avoid the "location highlighting bug" part of this issue in https://github.com/ocaml/ocaml/issues/11899#issuecomment-1383872695 that avoid the complexities of unicode (I was already pretty much convinced but I hope you two had fun):

- Using an on-line / overlay notification when supported,
- disabling the visual markers when using non-trivial characters otherwise

My gut feeling would be to stick to the current byte-reporting format for now, but it looks like this is a really exciting question to discuss further. Should I open separate issues for my two proposal above?

**Comment by hhugo** [comment]
> My understanding here is that most tools are still using "number of bytes" (and assuming UTF8 encoding)

I think it would be good to list tools you have in mind here, mostly for documentation purposes and help devs working on this in the future to evaluate the tradeoffs.

**Comment by gasche** [comment]
This was purely from memory of previous discussions and I have no idea, sorry. I'm afraid I'm not the right person to discuss the better way to indicate column numbers in modern editors.

**Comment by dbuenzli** [comment]
> … I hope you two had fun):

I like where @Octachron goes when you try to corner him, his last suggestion (tokens) is interesting. Note that in some writing systems like chinese there's no proper notion of token, but there is certainly one in OCaml.

> My gut feeling would be to stick to the current byte-reporting format for now, 

I agree but I still think that it should be properly documented (that's what #8978 was for) and explained so that external tooling knows what it gets, how it needs to interpret it and what happened when it will change (if it once does).

> but it looks like this is a really exciting question to discuss further. Should I open separate issues for my two proposal above?

I don't think this issue will be useful for acting. I support that move.

> I'm not the right person to discuss the better way to indicate column numbers in modern editors.

@rgrinberg and other merliners like @voodoos likely know better.

**Comment by gasche** [comment]
I created #11917 and #11918 meant as "actionable" issues for how to adapt our location-highlighting approaches, for rich terminals and for dumb terminals. I also reopened #8978 ("document the error location format") for the what-do-those-numbers-mean question. I think that this covers all two aspects of the discussion (numbers + highlighting) in a way that we could realistically do in the close/medium-term future. Is it okay to close the present issue?
